staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -212,63 +231,79 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-		case <-s.ServerShuttingDownCh:
-			// the server has signaled that it is shutting down (not accepting
-			// any new request), all active watch request(s) should return
-			// immediately here. The WithWatchTerminationDuringShutdown server
-			// filter will ensure that the response to the client is rate
-			// limited in order to avoid any thundering herd issue when the
-			// client(s) try to reestablish the WATCH on the other
-			// available apiserver instance(s).
-			return
-		case <-done:
-			return
-		case <-timeoutCh:
-			return
-		case event, ok := <-ch:
-			if !ok {
-				// End of results.
+	serveFunc := func() {
+		if s.MemoryAllocator != nil {
+			defer runtime.AllocatorPool.Put(s.MemoryAllocator)
+		}
+		defer s.Watching.Stop()
+		defer cleanup()
+
+		watchEncoder := newWatchEncoder(req.Context(), kind, s.EmbeddedEncoder, s.Encoder, framer)
+		ch := s.Watching.ResultChan()
+		done := req.Context().Done()
+
+		for {
+			select {
+			case <-s.ServerShuttingDownCh:
+				// the server has signaled that it is shutting down (not accepting
+				// any new request), all active watch request(s) should return
+				// immediately here. The WithWatchTerminationDuringShutdown server
+				// filter will ensure that the response to the client is rate
+				// limited in order to avoid any thundering herd issue when the
+				// client(s) try to reestablish the WATCH on the other
+				// available apiserver instance(s).
 				return
-			}
-			metrics.WatchEvents.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Inc()
-			isWatchListLatencyRecordingRequired := shouldRecordWatchListLatency(event)
-
-			if err := watchEncoder.Encode(event); err != nil {
-				utilruntime.HandleError(err)
-				// client disconnect.
+			case <-done:
 				return
-			}
+			case <-timeoutCh:
+				return
+			case event, ok := <-ch:
+				if !ok {
+					// End of results.
+					return
+				}
+				metrics.WatchEvents.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Inc()
+				isWatchListLatencyRecordingRequired := shouldRecordWatchListLatency(event)
 
-			if len(ch) == 0 {
-				flusher.Flush()
-			}
-			if isWatchListLatencyRecordingRequired {
-				metrics.RecordWatchListLatency(req.Context(), s.Scope.Resource, s.metricsScope)
+				if err := watchEncoder.Encode(event); err != nil {
+					utilruntime.HandleError(err)
+					// client disconnect.
+					return
+				}
+
+				if len(ch) == 0 {
+					flusher.Flush()
+				}
+				if isWatchListLatencyRecordingRequired {
+					metrics.RecordWatchListLatency(req.Context(), s.Scope.Resource, s.metricsScope)
+				}
 			}
 		}
 	}
+
+	metrics.RecordLongRunning(req, requestInfo, metrics.APIServerComponent, func() {
+		serveFunc()
+	})
 }
 
 // HandleWS implements a websocket handler.
 func (s *WatchServer) HandleWS(ws *websocket.Conn) {
 	defer ws.Close()
 	done := make(chan struct{})
+	// ensure the connection times out
+	timeoutCh, cleanup := s.TimeoutFactory.TimeoutCh()
```

Comment:
Same for this one - move it together to the PR that will only be touching timeouts.

Reply:
Done in #122060